api docs: Start exhaustively documenting our API.

eeshangarg opened the pull request:
@showell, @timabbott: I would really appreciate some feedback so that I can go ahead and document all of our API. Once I have a go-ahead from you, I just want to rinse and repeat the same process outlined in this PR for every API endpoint that we have till everything follows a macro-based structure similar to our webhooks' docs. Another side goal is to develop a way to test all Python code examples! :)

Also, I used [Slack's Web API docs](https://api.slack.com/methods/api.test) for a little bit of inspiration! Thanks! :)

showell commented on templates/zerver/help/include/successful-api-json-response.md:
````diff
@@ -0,0 +1,9 @@
+A typical successful JSON response may look like:
+
+```
+{
````
How universal is the actual response?  We do have many API calls that aren't dealing with a message.

eeshangarg replied:
@showell: From what I could tell (I could be wrong, but I'm hoping I won't be wrong enough for these macros to be useless beyond these two cases), a success response in a case where the only information that's returned is *just* the information that signals success and nothing beyond that, the message should be universal across any such calls. Again, I could run into trouble with that as I start documenting endpoints that have never been advertised.

I guess in other words, any call that returns data that's not just an indicator of success would be different, so you are right, its not wholly universal, but I judged it to be universal enough to warrant a separate macro. :)

showell replied:
I'd do one of the following:
* make the macro just be the wording part so that's consistent
* make the macro have a more specific name

I'm more inclined to the former, since I think it's good to have accurate responses when we can, especially if we can eventually have automated testing of the examples.

eeshangarg replied:
I'm leaning towards the latter instead because I don't see much of a variation in accuracy when it comes to at least these two cases and I bet this isn't the only instance that the response is exactly the same! Actually, I'll document a few more endpoints and I'll go with the first one if I find that that the universality of the response is less pervasive than I initially anticipated. :)

eeshangarg commented:
@showell: Updated, we now:

* Render the `Arguments` table from a JSON file.
* Also, I renamed the macro to be more specific for now!

The solution was not simple! Thanks so much for your guidance and suggestions! I would love your feedback before I move forward and work on this any further! Looking forward to it! Thanks! :)

showell commented on zerver/lib/bugdown/api_arguments_table_generator.py:
```diff
@@ -0,0 +1,108 @@
+                        break
+
+                    line_split = REGEXP.split(line, maxsplit=0)
+                    text[0] = line_split[0] + text[0]
```
I would add some comments here explaining what's going on, and possibly extract some local vars for clarity sake.

showell commented:
Looks great so far!  The markdown piece seems pretty clean to me, although you may want to comment fairly liberally here.

eeshangarg commented:
@showell: Updated! I have a few things I'd like to mention.

* I made a few comments to clarify some of the code in the new Markdown extension and also made the code a bit simpler than before.
* I documented the `GET /api/v1/streams` endpoint.
* Note that the JavaScript examples for the get-streams endpoint are missing an example for how to pass one of the documented query parameters. This is because `zulip-js` does not export a generic utility to interface with the API on a lower level (such as `call-endpoint` in our Python API), nor does the `streams.retrieve` function take any arguments. Should I maybe start committing code that facilitates said low-level interfacing right now and maybe ask the `zulip-js` maintainer to release it so that we can start documenting everything in JavaScript? Or would it better to document the whole API in `curl` and `Python` and then worry about JavaScript?

Also, I just removed the `WIP` label from the PR, so unless you have more feedback for the earlier commits, we could potentially start merging this incrementally! :) I have a feeling this PR might get a bit cluttered as I work on this further! :) Let me know what you think! Thanks!

showell commented:
For `zulip-js`, you can make the call based on how quickly we can make zulip-js more similar to Python.  It's not the end of the world if the JS stuff is a bit in flux for the short term, since hardly anybody uses it yet, and the cutting edge folks may just be reading the code.

showell commented:
Merged, thanks @eeshangarg!
